File: 4-constrained-optimization/matlab/ex_barrier.m
% ex_barrier.m
% example of a barrier method for constrained optimization
% [reference] Section 16.2.1 in LNO
% [course] Session 12 - Constrained Optimization (4) and Derivative-free
% Optimization (1)
close all; clear; clc

% problem functions (Example 16.1 in LNO)
f = @(x) x(1) - 2*x(2);
g = @(x) [1 + x(1) - x(2)^2; x(2)];

% initial strictly feasible point
x = [0.5;0.5];

% initial barrier parameter
mu = 1;

% barrier term
phi = @(x) -sum(log(g(x)));

% barrier function
beta = @(x,mu) f(x) + mu*phi(x);

% fminunc options
OPTIONS = optimoptions('fminunc');
OPTIONS.Display = 'none';
OPTIONS.StepTolerance = 0;
OPTIONS.OptimalityTolerance = 0;
OPTIONS.FunctionTolerance = 0;

% go through each iteration
for iter = 1:7

    % solve the unconstrained optimization problem
    x = fminunc(@(x) beta(x,mu),x,OPTIONS);

    % display stuff
    disp_helper("--- iteration",iter,[])
    disp_helper("mu",mu,7)
    disp_helper("x",x,7)

    % update mu
    mu = mu/10;

end

%--------------------------------------------------------------------------
% function to make it easier to display things in the command window
function disp_helper(name,number,n)

% default value of the number of digits
if isempty(n)
    n = 5;
end

% form string
str = strcat(string(name)," = ",mat2str(round(number,n)));

% display string
disp(str)

end

File: 4-constrained-optimization/python/ex_barrier.py
import numpy as np 
from scipy.optimize import minimize

def disp_helper(name,number,n):
    
    # default value of the number of digits
    if len(n) == 0:
        n = [5]
        
    # form string
    string = name + ' = {}'
    
    # display string
    print(string.format(number))
    

# problem functions
f = lambda x: x[0] -2*x[1]
g = lambda x: np.array([1 + x[0]-x[1]**2,x[1]])

# initial strictly feasible point
x = np.array([0.5,0.5])

# initial barrier parameter
mu = 1 

# barrier term
phi = lambda x: -np.sum(np.log(g(x)))

# barrier function
beta = lambda x,mu: f(x) + mu*phi(x)

# problem options
options={'xatol':0, 'fatol':0,'disp': False}

# go through each iteration
for i in range(9):
    
    res = minimize(beta,args = (mu),x0 = x,method = 'Nelder-Mead',options = options)
    
    # extract results
    x = res.x
    
    disp_helper('---iteration',i,[])
    disp_helper('mu',mu,[10])
    disp_helper('x',x,[10])
    print('')
    
    # update mu
    mu = mu/10

